### ISSUE TITLE
Automatically add tags based on library calls in post [FR]

### ISSUE THREAD

**Issue by brshallo** [rstudio/blogdown#647]
Would be nice if had easy way to set tags based on the library calls used in that post. Something like:

``` r
library(stringr)
library(magrittr)

tag_libs <- function(file_path, include_explicit = TRUE){
  
  file <- readr::read_lines(file_path)
  
  lib_calls <- str_extract_all(file, "(?<=library\\().+(?=\\))", simplify = TRUE)
  
  if(include_explicit){
    explicit_calls <- str_extract_all(file, "[[:alnum:]|\\.]+(?=::)", simplify = TRUE)
  } else {
    explicit_calls <- NULL
  }
  
  str_remove_all(lib_calls, '"') %>% 
    c(explicit_calls) %>% 
    str_subset(pattern = ".+") %>%
    sort() %>% 
    unique() %>% 
    str_flatten("\n  - ") %>% 
    str_c("  - ", .) %>% 
    cat()
}

# Example from one of my blogs
file_path <- "https://raw.githubusercontent.com/brshallo/brshallo/master/content/post/2021-03-18-intuition-on-uncertainty-of-predictions-introduction-to-prediction-intervals.html"

tag_libs(file_path)
#>   - AmesHousing
#>   - forcats
#>   - ggplot2
#>   - gt
#>   - pander
#>   - parsnip
#>   - scales
#>   - tidymodels
#>   - tidyverse
#>   - workflows
#>   - yardstick
```

<sup>Created on 2021-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Ideally set-up so may run without hassle of typing out `filepath`. For example could...

* be selected via an option when calling the "Update Metadata" or other add-in function.
* set in the YAML as an option under "tags" or under "output" --> "blogdown::html_page".

## Motivation

I can relate to @yihui comment https://github.com/rstudio/blogdown/issues/172#issuecomment-319569278

> Personally I don't even use categories or tags in my blogs since I'm lazy and I don't think my blogs are so big and complicated that I need taxonomies to manage them.

The tags I've self-set on posts have mostly ended-up being useless or essentially duplicates of my Category labels. I thought the option to use the library's called within a post would be a more natural way of setting/navigating tags on my rstats posts.

**Comment by yihui** [comment]
I think we can implement this in the "Update Metadata" addin. Thanks for the suggestion!

**Comment by brshallo** [comment]
Sounds great! 

(Thinking with YAML header approach would be to have a `blogdown::tag_libs` like function/setting that could be added -- e.g. useful if wanted to make this the default behavior via the template used on posts.)

**Comment by brshallo** [comment]
FYI, added the function above to a package [brshallo/funspotter](https://github.com/brshallo/funspotr) (written to *approximately* parse out functions & packages from self-contained R & Rmd files), where the example would be written as:

```r
funspotr::spot_pkgs(
  file_path = "https://raw.githubusercontent.com/brshallo/brshallo/master/content/post/2021-03-18-intuition-on-uncertainty-of-predictions-introduction-to-prediction-intervals.html",
  as_yaml_tags = TRUE
)
```

Though, as I mention there, something like the approach `renv::dependencies()` would be better than the quick regex's I use in funspotr.

**Comment by brshallo** [comment]
Would love to have this evaluated when knitting though, something like...

```r
---
title: This is a post
author: brshallo
date: '2022-02-11'
tags: "`r funspotr::spot_pkgs(rstudioapi::getSourceEditorContext()$path, as_yaml_tags = TRUE)`"
slug: this-is-a-post
---
```

So far unsuccessful with this or some other approaches I've tried -- opened SO question [Programmatically add tags to yaml header during knitting](https://stackoverflow.com/questions/71085925/programmatically-add-tags-to-yaml-header-during-knitting).

**Comment by yihui** [comment]
`rstudioapi::getSourceEditorContext()` only works when it is executed interactively inside the RStudio IDE. For `.Rmd` posts, they are knitted in new R sessions, and `rstudioapi::getSourceEditorContext()` won't work. In this case, I guess the content of the current file can be obtained from `xfun::read_utf8(knitr::current_input())`.

**Comment by brshallo** [comment]
Thanks for the note for a more general solution outside of interactive environments! 
funspotr expects a filepath so in this particular case would just input `knitr::current_input(dir = TRUE)`.

Still haven't figured out how to pass inline R code to `tags` in YAML header (per SO thread linked to above) so this still doesn't work:
```
tags: "`r funspotr::spot_pkgs(knitr::current_input(dir = TRUE), as_yaml_tags = TRUE)`"
```

**Comment by yihui** [comment]
Your problem is that all of your tags are quoted inside double quotes. That is, you were generating something like this:

```
tags: "
- AmesHousing
- forcats
- ...
"
```

which is not an array in YAML.

To generate a valid YAML array, you could use the alternative syntax `[]`, e.g.,

```
tags: ["`r paste(head(letters), collapse = '", "')`"]
```

which generates:

```
tags: ["a", "b", "c", "d", "e", "f"]
```

Note the hack `collapse = '", "'`: since there already exists a pair of double quotes outside the R expression, you should only generate the part `a", "b", "c", "d", "e", "f` from the R expression.

**Comment by brshallo** [comment]
Thanks! This works great!

```
tags: ["`r paste(funspotr::spot_pkgs(rstudioapi::getSourceEditorContext()$path), collapse = '", "')`"]
```

However going back to the general solution, `knitr::current_input()` output actually seems to be empty when in inline code in the YAML header of a *blogdown* post. Here's an example where I try and have the file name be in both the title and also the prefix of a tag (just to show 2 examples of it being empty):

```
---
title: "`r knitr::current_input()`"
author: Bryan Shalloway
date: '2022-01-18'
tags: ["`r paste0(knitr::current_input(), '__testing')`"]
...
---
```

![image](https://user-images.githubusercontent.com/30487784/154444166-3e3e34f5-6309-4f49-8bb2-9b92892e90f8.png)

Can see title is empty and suffix of tag (where filename should be) is missing. Note that `knitr::current_input()` still works if in a code chunk in the file:

![image](https://user-images.githubusercontent.com/30487784/154444950-e6abdeea-1c8b-4aa8-b76b-a4544ba8b5a6.png)


I confirmed that on a regular `rmarkdown::html_document` that `knitr::current_input()` works for me as expected when in a YAML header (i.e. as described on SO [here](https://stackoverflow.com/a/53516876/9059865) and [here](https://stackoverflow.com/a/31752708/9059865) where is used to populate Title in the YAML header).

Not sure if this may have to do with differences in how `blogdown::html_page` files are knitted...?

(Maybe should open a separate issue...?)

**Comment by yihui** [comment]
Currently it only works with the option `blogdown.method = 'markdown'` (https://bookdown.org/yihui/blogdown/global-options.html), or the file extension `.Rmarkdown`.

**Comment by brshallo** [comment]
Thank you for your patience.

Is there a reasonable work around that would allow for knitting .Rmd files to .html in blogdown *as well*?

---

I searched `knitr::current_input()` on Stack Overflow but the alternatives I found also didn't seem to work with blogdown Rmd's, as in this case the working directory at the point of evaluating inline code in the YAML header seems to be the root of the project whereas when evaluating code in the document the working directory is the location of the file -- which is different from other RMD types, where at both points (in-line YAML and in-document evaluation) the default working directory is generally the location of the file.

P.s. May not be worth effort if solution is really convoluted or limited as I think *most* of the time I would still just specify tags interactively like: https://twitter.com/brshallo/status/1493143780886597634?s=20&t=NcO3SrA9qrVYD8uHhYEwKQ

I just like the idea of having the *ability* to simply paste in a function that could handle it.
